### PR TITLE
Lazy Sequence Decoder

### PR DESCRIPTION
**Pull request by josh**
Initial implementation at a sequence decoder API as described in #15.

### CODE REVIEW

**Comment by josh** [comment]
```swift
let decoder = CSVDecoder(configuration: ...)
for row in decoder.lazy(from: url) {
    let student = try row.decode(Student.self)
    // Do something
}
```

☝️ This API should work.


```swift
let decoder = CSVDecoder(configuration: ...)
let sequence = decoder.lazy(from: url)
guard let rowA = sequence.next(),
      let rowB = sequence.next() else { ... }
let studentB = try rowB.decode(Student.self)
let studentA = try rowA.decode(Student.self)
```

☝️ Works with `bufferingStrategy: .keepAll ` but fails on `.sequential`. I think this could be considered acceptable.

> We should be clear in the documentation that that is the case and decide what to do with some Sequence properties such as underestimatedCount

I think we can return the current number of decoded rows (`currentIndex`). It's supposed to be `O(1)` so we should probably not do a full pass.

> Take to heart what itaiferber said about keeping decoding containers alive. They are intended for scoped in-order use and they all reuse the same ShadowDecoder.Source.

No decoding containers are re-used here.

File: sources/declarative/decodable/DecoderSequence.swift [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+            return nil
+        }
+
+        defer { self.currentIndex += 1 }
+        let decoder = ShadowDecoder(source: source, codingPath: [IndexKey(self.currentIndex)])
+        return CSVRowDecoder(decoder: decoder)
+    }
+}
+
+public struct CSVRowDecoder {
```
Comment (dehesa) [inline sources/declarative/decodable/DecoderSequence.swift:21]: Same thing as with the sequence. I would prefer to put this in the same namespace. In this case:
```swift
extension CSVDecoder.LazySequence {
    public struct Row { ... }
}
```

File: sources/declarative/decodable/DecoderSequence.swift [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    /// The representation of the decoding process point-in-time.
+    let decoder: ShadowDecoder
```
Comment (dehesa) [inline sources/declarative/decodable/DecoderSequence.swift:23]: Let's make it `private` and then have a `fileprivate` initializer to pass the `decoder`.

File: sources/declarative/decodable/DecoderSequence.swift [hunk 3]
```diff
@@ -0,0 +1,31 @@
+open class CSVDecoderSequence: IteratorProtocol, Sequence {
+    private let source: ShadowDecoder.Source
+    private var currentIndex: Int = 0
```
Comment (dehesa) [inline sources/declarative/decodable/DecoderSequence.swift:3]: Should we expose `currentIndex`? (`private(set) public var`) Do you see any value on this?

Comment (josh) [reply sources/declarative/decodable/DecoderSequence.swift:3]: Probably not. If you wanted the row index, you could use `enumerated()`.

**Comment by josh** [comment]
Thanks for the review! Pushed 16ed7e968e3d06e68ec99c92bb8c4a8d06dc6ceb

**Comment by dehesa** [comment]
Thank you for the PR @josh. Great stuff as always.

**Pull request merged**